File: ocaml-examples-3.12/simples/date.pl
sub norm_date 
{
    my $d = @_[0] ;
    chomp $d ;
    if ( $d =~ /bientot/i ) {return($d);}
    if ( $d =~ /bientôt/i ) {return($d);}
    if ( $d =~ /interdit/i ) {return($d);}
    my $j, $m, $a, $jj, $mm, $aa, $sep ;
    if ( $d =~ m!^(\d{1,2})([ :/\.\-])(\d{1,2})\2(\d{1,4})$! ) {
        $j = &norm_chiffre($1) ; $m = &norm_chiffre($3);
        $a = $4 ;
        if ($a>2000) { $a = $a - 2000 ; }
        $a = &norm_chiffre ($a) ;
        return (join ('/', ($j,$m,$a) ) ) ;
    }
    return ($d) ;
}
sub norm_chiffre {
    my $c = shift ;
    if (length($c)<2) {$c = "0".$c;}
    $c;
}

File: ocaml-examples-3.12/simples/date.ml
(***********************************************************************)
(*                                                                     *)
(*                        Caml examples                                *)
(*                                                                     *)
(*            Pierre Weis                                              *)
(*                                                                     *)
(*                        INRIA Rocquencourt                           *)
(*                                                                     *)
(*  Copyright (c) 1994-2011, INRIA                                     *)
(*  All rights reserved.                                               *)
(*                                                                     *)
(*  Distributed under the BSD license.                                 *)
(*                                                                     *)
(***********************************************************************)

(* $Id: date.ml,v 1.3 2011-08-08 19:31:17 weis Exp $ *)

let norm_chiffre c =
  if c >= 10 then string_of_int c else "0" ^ string_of_int c
;;

(* Should be a full regular expression matching. *)
let ( =~ ) s pat =
  let lp = String.length pat and ls = String.length s in
  lp <= ls && pat = String.sub s 0 lp
;;

let norm_date d =
  if d =~ "bientot" ||
     d =~ "bientôt" ||
     d =~ "interdit" then d else
  Scanf.sscanf d
   " %2u %_1[:/.-] %2u %_1[:/.-] %4u  %!"
   (fun j m a ->
     let a = if a > 2000 then a - 2000 else a in
     String.concat "/" (List.map norm_chiffre [j; m; a]))
;;
